File: shared/types/session.ts
import type { PCRStateEnvelope } from './pcr';

export interface TranscriptSegment {
  text: string;
  start_time: number;
  end_time: number;
  timestamp: string;
}

export interface CorrectionEvent {
  field_name: string;
  old_value: unknown;
  new_value: unknown;
  timestamp: string;
}

export interface GapItem {
  field_name: string;
  usage: 'mandatory' | 'required' | 'recommended';
  section: string;
  description: string;
  prompt: string;
  priority: number;
}

export interface GapDetectionResult {
  missing_mandatory: GapItem[];
  missing_required: GapItem[];
  missing_recommended: GapItem[];
  suggested_prompts: string[];
  total_gaps: number;
}

export interface SessionResponse {
  session_id: string;
  incident_id: string | null;
  created_at: string;
  status: 'active' | 'finalized' | 'archived';
  pcr_state: PCRStateEnvelope;
}

File: backend/app/schemas/session.py
"""Session request/response models."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.schemas.pcr import PCRStateEnvelope


class SessionCreateRequest(BaseModel):
    incident_id: Optional[str] = None


class SessionResponse(BaseModel):
    session_id: str
    incident_id: Optional[str] = None
    created_at: datetime
    status: str = "active"
    pcr_state: PCRStateEnvelope


class TranscriptSegment(BaseModel):
    """A single transcript segment with timestamp."""

    text: str
    start_time: Optional[float] = None
    end_time: Optional[float] = None
    timestamp: datetime


class CorrectionEvent(BaseModel):
    """Record of a correction applied to the PCR."""

    utterance: str
    field: str
    old_value: Optional[str] = None
    new_value: str
    timestamp: datetime


class SessionDetailResponse(BaseModel):
    session_id: str
    incident_id: Optional[str] = None
    created_at: datetime
    status: str
    pcr_state: PCRStateEnvelope
    transcript_history: list[TranscriptSegment] = Field(default_factory=list)
    correction_history: list[CorrectionEvent] = Field(default_factory=list)
